### PR TITLE
refactor(onboarding): consolidate onboarding flow logic.

### PR DESCRIPTION
**Pull request by jh2oman**
### Description

This solves a few problems with the current onboarding flow logic:
1. The logic for which screen to go to next in onboarding is currently spread out throughout the different onboarding screens. This makes it difficult to see in-code how the onboarding flow actually works with its different branching options. With this refactor all that logic would be in one place
2. The logic for the 'Step X of Y' is currently totally separate from the logic of which screens come after one another in onboarding. This duplication of logic is confusing and messy. With this refactor we have a traversable graph that can be used to calculate the steps without separate logic that must change every time a screen is added.

### Test plan

1. More unit test coverage than there was before for the onboarding flow
2. Unit tests of the Onboarding directed graph ensuring that counters always increment and the graphs always end up at Home.

### Related issues

https://linear.app/valora/issue/ACT-597/refactor-onboarding-screen-logic

### Backwards compatibility

yes

### CODE REVIEW

File: src/onboarding/map.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+/**
+ * Traverses through the directed graph of onboarding navigate, navigateClearingStack, and navigateHome calls
+ * and count the number of screens until the given screen and the total number
+ */
+export function getStepValues(screen: Screens) {
+  let stepCounter = 0
+  let totalCounter = 0
+  let reachedScreeen = false
+  let currentScreen: Screens = Screens.NameAndPicture
```
Comment (cajubelt) [inline src/onboarding/map.ts:34]: would be nice if `Screens.NameAndPicture` was a constant somewhere named FIRST_ONBOARDING_SCREEN or something

File: src/onboarding/map.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    if (!reachedScreeen) {
+      stepCounter++
+    }
+    if (currentScreen === screen) {
+      reachedScreeen = true
+    }
+  }
+
+  const navigateHome = () => {
+    currentScreen = Screens.WalletHome
```
Comment (cajubelt) [inline src/onboarding/map.ts:48]: seems like there could be a constant for `Screens.WalletHome` like AFTER_ONBOARDING_SCREEN . This could actually change in the future, for example with the "choose your adventure" experience

File: src/onboarding/map.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    navigate: NavigationService.navigate,
+    popToScreen: NavigationService.popToScreen,
+    navigateHome: NavigationService.navigateHome,
+    navigateClearingStack: NavigationService.navigateClearingStack,
+    dispatch: store.dispatch,
+  })
+  onBoardingMap[screen]?.next()
+}
+
+function _getOnboardingMap({
```
Comment (cajubelt) [inline src/onboarding/map.ts:83]: this is really interesting! I love how it uses dependency injection to make it possible to just traverse the graph without creating a bunch of side effects.

I almost want to take it a step further where it takes all this state as a parameter-- or maybe uses some TS magic to only require the state that's needed from the start screen you're at??-- and returns a screen you need to go to, plus some way of telling what side effects need to take place if you're going to that screen in a "real" onboarding navigation flow. it could be just a function like "doSideEffects". the `navigateClearingStack` use case is annoying, but maybe there's some way of taking what that does and separating out the effectful part into the `doSideEffects` piece.

It's already in pretty good shape so it's hard to say what's an improvement about my idea exactly. Taking state as a parameter makes it a pure function, so that's good. I guess separating the side effects out into a function on the return value makes it feel more fair to use it for `getStepValues`. 

I think my main feeling is that this isn't really a "map"; it's more like a state-dependent set of edges between screens that have a start screen, an end screen, and a side effect for the transition between the screens (which you may or may not want to use!).

**Review by cajubelt — COMMENTED** [review]
really cool!

File: src/onboarding/success/OnboardingSuccessScreen.tsx [hunk 4]
```diff
@@ -4,14 +4,20 @@ import { Image, StyleSheet, Text, View } from 'react-native'
...
 import colors from 'src/styles/colors'
 import fontStyles from 'src/styles/fonts'
 import { Spacing } from 'src/styles/styles'
 
 function OnboardingSuccessScreen() {
+  const onboardingProps = useTypedSelector(onboardingPropsSelector)
   useEffect(() => {
-    const timeout = setTimeout(() => navigateHome(), 3000)
+    const timeout = setTimeout(
+      () => goToNextOnboardingScreen(Screens.VerificationStartScreen, onboardingProps),
```
Comment (cajubelt) [inline src/onboarding/success/OnboardingSuccessScreen.tsx:18]: i find this a bit confusing. the intuitive screen parameter seems like it would be Screens.OnboardingSuccessScreen. can you explain why it's not?

File: src/verify/VerificationStartScreen.tsx [hunk 5]
```diff
@@ -81,7 +82,11 @@ function VerificationStartScreen({
   const cachedNumber = useSelector(e164NumberSelector)
   const cachedCountryCallingCode = useSelector(defaultCountryCodeSelector)
   const walletAddress = useSelector(walletAddressSelector)
-  const { step, totalSteps } = useSelector(registrationStepsSelector)
+  const onboardingProps = useTypedSelector(onboardingPropsSelector)
+  const { step, totalSteps } = getOnboardingStepValues(
+    Screens.VerificationStartScreen,
+    onboardingProps
+  )
```
Comment (cajubelt) [inline src/verify/VerificationStartScreen.tsx:89]: looks like there is still some bespoke logic for determining the next onboarding screen to show from this one-- not ideal

Comment (cajubelt) [reply src/verify/VerificationStartScreen.tsx:89]: it seems like the problem is that there are 2 screens you might want to show next-- OnboardingSuccessScreen if they successfully verify, or home if they skip. Two ideas for handling this branching:
- treat "home" as the next screen from this one, and pass a callback to OnboardingSuccessScreen as a prop that takes it to the "next screen in onboarding" (always home)
- put the branching logic inside `goToNextOnboardingScreen`, and choose the next screen by checking whether a user has verified their PN

**Review by cajubelt — COMMENTED** [review]
still need to review onboarding/steps.ts and steps.test.ts-- skipped for now since they look similar to my last review

File: src/onboarding/steps.ts [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  supportedBiometryTypeSelector,
+} from 'src/app/selectors'
+import { setHasSeenVerificationNux } from 'src/identity/actions'
+import * as NavigationService from 'src/navigator/NavigationService'
+import { Screens } from 'src/navigator/Screens'
+import { RootState } from 'src/redux/reducers'
+import { store } from 'src/redux/store'
+
+export const FIRST_ONBOARDING_SCREEN = Screens.NameAndPicture
+export const END_OF_ONBOARDING_SCREEN = Screens.WalletHome
```
Comment (cajubelt) [inline src/onboarding/steps.ts:19]: this might be a premature optimization, but in a future where we want to choose the first/last onboarding screen by the user's A/B test cohort, it would be more convenient to have these wrapped by getter functions, like `getFirstOnboardingScreen()`

Comment (cajubelt) [reply src/onboarding/steps.ts:19]: they can still just return a constant value for now. just thinking if they're exported and used elsewhere, it might be better to wrap them in getter functions

File: src/onboarding/steps.ts [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+interface NavigatorFunctions {
+  navigate: typeof NavigationService.navigate | ((screen: Screens) => void)
```
Comment (cajubelt) [inline src/onboarding/steps.ts:22]: i don't really get the OR types here. why is that needed?

File: src/onboarding/steps.ts [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+  let stepCounter = 1
+  let totalCounter = 1
+  let reachedScreeen = false
+  let currentScreen: Screens = FIRST_ONBOARDING_SCREEN
+
+  const navigate = (s: Screens) => {
+    totalCounter++
+    if (currentScreen === screen) {
+      reachedScreeen = true
+    }
```
Comment (cajubelt) [inline src/onboarding/steps.ts:75]:
```suggestion
    reachedScreen = reachedScreen || (currentScreen === screen)
```
code golf. very optional.

File: src/onboarding/steps.ts [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+ * Traverses through the directed graph of onboarding navigate, navigateClearingStack, and navigateHome calls
+ * and count the number of screens until the given screen and the total number
+ */
+export function getOnboardingStepValues(screen: Screens, onboardingProps: OnboardingProps) {
+  let stepCounter = 1
+  let totalCounter = 1
+  let reachedScreeen = false
+  let currentScreen: Screens = FIRST_ONBOARDING_SCREEN
+
+  const navigate = (s: Screens) => {
```
Comment (cajubelt) [inline src/onboarding/steps.ts:71]:
```suggestion
  const navigate = (s: Screens) => {
      // dummy navigation function to help determine what onboarding step the user is on, without triggering side effects like actually cycling them back through the first few onboarding screens
```

Comment (cajubelt) [reply src/onboarding/steps.ts:71]: also, I generally object to single-character variable names. `toScreen` could be more clear

File: src/onboarding/steps.ts [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+/**
+ * Traverses through the directed graph of onboarding navigate, navigateClearingStack, and navigateHome calls
+ * and count the number of screens until the given screen and the total number
+ */
+export function getOnboardingStepValues(screen: Screens, onboardingProps: OnboardingProps) {
+  let stepCounter = 1
+  let totalCounter = 1
+  let reachedScreeen = false
+  let currentScreen: Screens = FIRST_ONBOARDING_SCREEN
```
Comment (cajubelt) [inline src/onboarding/steps.ts:69]:
```suggestion
  let currentScreen: Screens = FIRST_ONBOARDING_SCREEN  // pointer that we will update when simulating navigation through the onboarding screens to calculate "step" and "totalSteps"
```

File: src/onboarding/steps.ts [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+/**
+ * Traverses through the directed graph of onboarding navigate, navigateClearingStack, and navigateHome calls
+ * and count the number of screens until the given screen and the total number
+ */
+export function getOnboardingStepValues(screen: Screens, onboardingProps: OnboardingProps) {
+  let stepCounter = 1
+  let totalCounter = 1
+  let reachedScreeen = false
```
Comment (cajubelt) [inline src/onboarding/steps.ts:68]:
```suggestion
  let reachedScreeen = false  // tracks whether we have reached the screen the user is on in onboarding, and we can stop incrementing stepCounter
```

File: src/onboarding/steps.ts [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+    numberAlreadyVerifiedCentrally,
+  }
+}
+
+/**
+ * Traverses through the directed graph of onboarding navigate, navigateClearingStack, and navigateHome calls
+ * and count the number of screens until the given screen and the total number
+ */
+export function getOnboardingStepValues(screen: Screens, onboardingProps: OnboardingProps) {
+  let stepCounter = 1
```
Comment (cajubelt) [inline src/onboarding/steps.ts:66]:
```suggestion
  let stepCounter = 1  // will increment this up to the onboarding step the user is on
```

File: src/onboarding/steps.ts [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+      reachedScreeen = true
+    }
+    if (!reachedScreeen) {
+      stepCounter++
+    }
+    currentScreen = s
+  }
+
+  const navigateHome = () => {
+    currentScreen = END_OF_ONBOARDING_SCREEN
```
Comment (cajubelt) [inline src/onboarding/steps.ts:83]:
```suggestion
    currentScreen = Screens.WalletHome
```

Comment (cajubelt) [reply src/onboarding/steps.ts:83]: if we ever updated END_OF_ONBOARDING_SCREEN, we wouldn't want `navigateHome` to not take you home anymore

File: src/onboarding/steps.ts [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+  const { navigate, popToScreen, navigateHome, navigateClearingStack } = navigator
+  const {
+    recoveringFromStoreWipe,
+    choseToRestoreAccount,
+    supportedBiometryType,
+    skipVerification,
+    numberAlreadyVerifiedCentrally,
+  } = props
+
+  const onboardingScreens: Partial<{
```
Comment (cajubelt) [inline src/onboarding/steps.ts:140]: seems like this could be a switch

**Review by cajubelt — COMMENTED** [review]
ran out of time before i could actually check the logic for each of the screen transitions, which is probly pretty important. would be great if another reviewer could look at that, or if you could self-review.

**Comment by cajubelt** [comment]
also one thing that concerns me a little is that the user's state could potentially change between when they start onboarding and when they finish, which could introduce a bug like a step getting added to the number of total steps, or even a cycle. not sure what the best way of dealing with that is, though.. maybe just making sure it's covered in integration tests?

File: src/onboarding/steps.test.ts [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+        supportedBiometryType: [null, BIOMETRY_TYPE.FACE_ID],
+        skipVerification: [true, false],
+        numberAlreadyVerifiedCentrally: [true, false],
+      }
+
+      options.recoveringFromStoreWipe.forEach((recoveringFromStoreWipe) => {
+        options.choseToRestoreAccount.forEach((choseToRestoreAccount) => {
+          options.supportedBiometryType.forEach((supportedBiometryType) => {
+            options.skipVerification.forEach((skipVerification) => {
+              options.numberAlreadyVerifiedCentrally.forEach((numberAlreadyVerifiedCentrally) => {
```
Comment (cajubelt) [inline src/onboarding/steps.test.ts:41]: i think this could be a little cleaner if it were refactored as follows:
- take "options" out of the test body
- prep a list of cases to check by iterating over the options like you do here, and just pushing to an array of cases to check
- pass that list of cases to check into an it.each

you could also use [this SO post](https://stackoverflow.com/questions/12303989/cartesian-product-of-multiple-arrays-in-javascript) to prep the cases to check with a 1-line helper, but that's probably overkill

if nothing else, it would be nice to see this test written without all the nested indentation

File: src/onboarding/steps.test.ts [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+                // Initiate Test Variables
+                const onboardingProps = {
+                  recoveringFromStoreWipe,
+                  choseToRestoreAccount,
+                  supportedBiometryType,
+                  skipVerification,
+                  numberAlreadyVerifiedCentrally,
+                }
+                let currentScreen: Screens = FIRST_ONBOARDING_SCREEN
```
Comment (cajubelt) [inline src/onboarding/steps.test.ts:50]: this test looks like it largely duplicates the implementation of getOnboardingStepValues, and then checks that the answer is the same at the end. This seems too circular to me. Also, it's complex enough that there could easily be bugs in the test itself.

Comment (cajubelt) [reply src/onboarding/steps.test.ts:50]: one way you could write this test is select a subset of cases for your test and hardcoding the expected values (after checking them manually). But computing them on the fly like this with essentially the same code is going to make the test less reliable, harder to maintain, and less useful overall

**Review by cajubelt — CHANGES_REQUESTED** [review]
concerned about the complexity of one of the tests

File: src/onboarding/steps.test.ts [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+
+                const navigate = (s: Screens) => {
+                  currentScreen = s
+                  stepCounter++
+                }
+                const navigateHome = () => {
+                  currentScreen = END_OF_ONBOARDING_SCREEN
+                }
+
+                const { totalSteps } = getOnboardingStepValues(
```
Comment (cajubelt) [inline src/onboarding/steps.test.ts:61]: it looks like all your tests work with the same screen parameter-- would be good to get better coverage on that. the highest impact cases to cover are probably the first screen, the last screen, and some screen in the middle.

Comment (jh2oman) [reply src/onboarding/steps.test.ts:61]: All of those screen parameters are tested below in the `expect(steps.totalSteps).toEqual(totalSteps)`, testing that throughout the flow the totalCounts are the same.

File: src/onboarding/steps.ts (hunk 7 above)
Comment (jh2oman) [reply src/onboarding/steps.ts:22]: No, the first type is needed for actually executing the navigation. The second type is needed for calculating the steps

File: src/verify/VerificationStartScreen.tsx (hunk 5 above)
Comment (jh2oman) [reply src/verify/VerificationStartScreen.tsx:89]: The way I think about it is VerificationStartScreen, VerificationInputScreen, and OnboardingSuccessScreen are all part of the Verification Step, and the logic for navigation between those screens should be handled internally since the logic is tightly coupled. 

The functions like `goToNextOnboardingScreen` should only be used to navigate between loosely coupled steps in the onboarding flow. Maybe I should change the name to `goToNextOnboardingStep` and create a new `OnboardingSteps` enum to capture the difference between screens and steps in this case.

File: src/onboarding/success/OnboardingSuccessScreen.tsx (hunk 4 above)
Comment (jh2oman) [reply src/onboarding/success/OnboardingSuccessScreen.tsx:18]: Its really the first screen in the current 'step'. I've adjusted the variable names to make this very clear.

File: src/onboarding/steps.test.ts [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+    ({ onboardingProps, screens, name }) => {
+      const expectedTotalSteps = screens.length
+      screens.forEach((screen, index) => {
+        const { totalSteps, step } = getOnboardingStepValues(screen, onboardingProps)
+        // Checking that the step number is correct
+        try {
+          expect(step).toEqual(index + 1)
+          expect(totalSteps).toEqual(expectedTotalSteps)
+        } catch (error) {
+          throw new Error(
```
Comment (cajubelt) [inline src/onboarding/steps.test.ts:81]: nit: jest already emits a similar error message

Comment (cajubelt) [reply src/onboarding/steps.test.ts:81]: maybe you were worried about figuring out which test failed. but you should be able to inject parameters from the test cases into the test description. (that being said, I tried adding "$name" to the end of the test description and it didn't work-- not sure why. maybe the jest version?)

https://jestjs.io/docs/26.x/api#testeachtablename-fn-timeout

File: src/onboarding/steps.test.ts [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+          // eslint-disable-next-line jest/no-conditional-expect
+          expect(navigateHome).toHaveBeenCalled()
+        } else {
+          try {
+            // eslint-disable-next-line jest/no-conditional-expect
+            expect(navigate).toHaveBeenCalledWith(screens[index + 1])
+          } catch {
+            // eslint-disable-next-line jest/no-conditional-expect
+            expect(navigateClearingStack).toHaveBeenCalledWith(screens[index + 1])
+          }
```
Comment (cajubelt) [inline src/onboarding/steps.test.ts:102]:
```suggestion
                expect(index === screens.length - 1 ? navigateHome : ((navigate as jest.MockedFunction<any>).mock.calls.length ? navigate : navigateClearingStack)).toHaveBeenCalled()
```
this is pretty ugly but I think it works

Comment (cajubelt) [reply src/onboarding/steps.test.ts:102]: and it lets us not disable linting

Comment (cajubelt) [reply src/onboarding/steps.test.ts:102]: oh just realized it means we can't use toHaveBeenCalledWith. Ugh. your original is probly better.

File: src/onboarding/steps.test.ts [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+      Screens.NameAndPicture,
+      Screens.PincodeSet,
+      Screens.EnableBiometry,
+      Screens.ImportWallet,
+      Screens.VerificationStartScreen,
+    ],
+    name: 'importWalletFlowEverythingEnabled',
+  }
+
+  it.each([
```
Comment (cajubelt) [inline src/onboarding/steps.test.ts:66]: the refactor looks very good!

File: src/onboarding/steps.ts [hunk 21]
```diff
@@ -63,12 +63,13 @@ export function onboardingPropsSelector(state: RootState): OnboardingProps {
  * and count the number of screens until the given screen and the total number
  */
 export function getOnboardingStepValues(screen: Screens, onboardingProps: OnboardingProps) {
-  let stepCounter = 1
+  let stepCounter = 1 // will increment this up to the onboarding step the user is on
   let totalCounter = 1
-  let reachedScreeen = false
-  let currentScreen: Screens = FIRST_ONBOARDING_SCREEN
+  let reachedScreeen = false // tracks whether we have reached the screen the user is on in onboarding, and we can stop incrementing stepCounter
```
Comment (cajubelt) [inline src/onboarding/steps.ts:68]:
```suggestion
  let reachedScreeen = false // tracks whether we have reached the step the user is on in onboarding, and we can stop incrementing stepCounter
```
since you clarified the screens vs steps thing, I realized this should be "step", not "screen"

**Review by cajubelt — APPROVED** [review]

File: src/onboarding/steps.ts [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+      navigateHome: NavigationService.navigateHome,
+      navigateClearingStack: NavigationService.navigateClearingStack,
+    },
+    dispatch: store.dispatch,
+    props: onboardingProps,
+  })
+  stepInfo?.next()
+}
+
+export function _getStepInfo({ firstScreenInStep, navigator, dispatch, props }: GetStepInfoProps) {
```
Comment (jophish) [inline src/onboarding/steps.ts:137]: I think a docstring here could be helpful, since this is now the source of truth for onboarding navigation logic. Probably would be useful to mention this.

File: src/onboarding/steps.ts [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+  const { navigate, popToScreen, navigateHome, navigateClearingStack } = navigator
+  const {
+    recoveringFromStoreWipe,
+    choseToRestoreAccount,
+    supportedBiometryType,
+    skipVerification,
+    numberAlreadyVerifiedCentrally,
+  } = props
+
+  switch (firstScreenInStep) {
```
Comment (jophish) [inline src/onboarding/steps.ts:147]: Should this have a default case which logs an error or something similar?

File: src/onboarding/steps.ts [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+            dispatch(setHasSeenVerificationNux(true))
+            navigateHome()
+          } else {
+            navigateClearingStack(Screens.VerificationStartScreen)
+          }
+        },
+      }
+    case Screens.EnableBiometry:
+      return {
+        next: () => {
```
Comment (jophish) [inline src/onboarding/steps.ts:180]: Why is the return signature of this function an object with a single property? Can we just return the `next` function directly?

File: src/onboarding/steps.ts [hunk 25]
```diff
@@ -0,0 +1,214 @@
...
+ * Traverses through the directed graph of onboarding navigate, navigateClearingStack, and navigateHome calls
+ * and count the number of screens until the given screen and the total number
+ */
+export function getOnboardingStepValues(screen: Screens, onboardingProps: OnboardingProps) {
+  let stepCounter = 1 // will increment this up to the onboarding step the user is on
+  let totalCounter = 1
+  let reachedScreeen = false // tracks whether we have reached the screen the user is on in onboarding, and we can stop incrementing stepCounter
+  let currentScreen: Screens = FIRST_ONBOARDING_SCREEN // pointer that we will update when simulating navigation through the onboarding screens to calculate "step" and "totalSteps"
+
+  const navigate = (s: Screens) => {
```
Comment (jophish) [inline src/onboarding/steps.ts:71]: nit: perhaps adjust naming on `navigate` and `navigateHome` to clarify that they're dummy functions? Without looking at the definitions, I would think that `navigate` and `navigateHome` actually do the thing they claim to.

File: src/onboarding/steps.ts [hunk 26]
```diff
@@ -0,0 +1,214 @@
...
+
+export interface OnboardingProps {
+  recoveringFromStoreWipe: boolean
+  choseToRestoreAccount: boolean | undefined
+  supportedBiometryType: BIOMETRY_TYPE | null
+  skipVerification: boolean
+  numberAlreadyVerifiedCentrally: boolean
+}
+
+export function onboardingPropsSelector(state: RootState): OnboardingProps {
```
Comment (jophish) [inline src/onboarding/steps.ts:45]: Maybe add a docstring explaining that any data that parameterizes onboarding screen logic must be included in this function's result?

**Review by jophish — COMMENTED** [review]
such a cool approach!!
